### PR TITLE
add git-sync-ssh secret template

### PR DESCRIPTION
**Pull request by romsharon98**
add git-sync-ssh secret as part of the chart in template.

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: chart/templates/secrets/git-ssh-key-secret.yaml [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+kind: Secret
+metadata:
+  labels:
+    release: {{ .Release.Name }}
+    chart: {{ .Chart.Name }}
+    heritage: {{ .Release.Service }}
+    {{- with .Values.labels }}
+      {{- toYaml . | nindent 4 }}
+    {{- end }}
+  name: airflow-ssh-secret
```
Comment (jedcunningham) [inline chart/templates/secrets/git-ssh-key-secret.yaml]: This name need to be updated to use `airflow.fullname`, like this: https://github.com/apache/airflow/blob/93ba263732a30ed51fdae0b2bff7cd959be115b8/chart/templates/secrets/webserver-secret-key-secret.yaml#L28

You'll also want to this secret to be used automatically (as in, don't require the user to configure the key name too). It'll be similar to this: https://github.com/apache/airflow/blob/93ba263732a30ed51fdae0b2bff7cd959be115b8/chart/templates/_helpers.yaml#L390-L392
You'll likely need a bit more logic where the key is mounted though.

File: chart/values.yaml [hunk 2]
```diff
@@ -2494,14 +2494,7 @@ dags:
     #
     # If you are using an ssh clone url, you can load
     # the ssh private key to a k8s secret like the one below
-    #   ---
```
Comment (jedcunningham) [inline chart/values.yaml]: Don't remove the example. Not everyone will want to include the key in their values file.

File: chart/templates/secrets/git-ssh-key-secret.yaml [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+data:
+  gitSshKey: {{ .Values.dags.gitSync.gitSshKey | quote }}
```
Comment (jedcunningham) [inline chart/templates/secrets/git-ssh-key-secret.yaml]: At least so far, we've accepted non-base64'd secret values, like webserver secret key. Is there a reason to require base64'd values for this?

e.g. https://github.com/apache/airflow/blob/93ba263732a30ed51fdae0b2bff7cd959be115b8/chart/templates/secrets/webserver-secret-key-secret.yaml#L40

**Review by jedcunningham — CHANGES_REQUESTED** [review]

File: chart/templates/secrets/git-ssh-key-secret.yaml (hunk 1 above)
Comment (romsharon98) [reply chart/templates/secrets/git-ssh-key-secret.yaml]: Thanks, change it like this given example

File: chart/values.yaml (hunk 2 above)
Comment (romsharon98) [reply chart/values.yaml]: reverted and add docs to explain my change

File: chart/templates/secrets/git-ssh-key-secret.yaml (hunk 3 above)
Comment (romsharon98) [reply chart/templates/secrets/git-ssh-key-secret.yaml]: I assume it from the [doc](https://github.com/apache/airflow/blob/main/docs/helm-chart/manage-dags-files.rst#mounting-dags-from-a-private-github-repo-using-git-sync-sidecar) but reverted it like you said

File: chart/values.schema.json [hunk 4]
```diff
@@ -7409,6 +7409,14 @@
                             ],
                             "default": null
                         },
+                        "gitSshKey": {
```
Comment (jedcunningham) [inline chart/values.schema.json]:
```suggestion
                        "sshKey": {
```

I think maybe `sshKey` is a better name for this?

File: chart/templates/secrets/git-ssh-key-secret.yaml (hunk 3 above)
Comment (jedcunningham) [reply chart/templates/secrets/git-ssh-key-secret.yaml]: This is still expecting a b64 string?

File: helm_tests/other/test_git_sync_scheduler.py [hunk 5]
```diff
@@ -211,6 +211,45 @@ def test_validate_if_ssh_params_are_added(self):
...
+    def test_validate_if_ssh_params_are_added_with_git_ssh_key(self):
+        docs = render_chart(
+            values={
+                "dags": {
+                    "gitSync": {
+                        "enabled": True,
+                        "containerName": "git-sync-test",
+                        "gitSshKey": "dummy-ssh-key",
+                        "knownHosts": None,
+                        "branch": "test-branch",
```
Comment (jedcunningham) [inline helm_tests/other/test_git_sync_scheduler.py]: We should either remove this or test it was set properly.

File: helm_tests/other/test_git_sync_scheduler.py [hunk 6]
```diff
@@ -211,6 +211,45 @@ def test_validate_if_ssh_params_are_added(self):
...
 
+    def test_validate_if_ssh_params_are_added_with_git_ssh_key(self):
+        docs = render_chart(
+            values={
+                "dags": {
+                    "gitSync": {
+                        "enabled": True,
+                        "containerName": "git-sync-test",
+                        "gitSshKey": "dummy-ssh-key",
+                        "knownHosts": None,
```
Comment (jedcunningham) [inline helm_tests/other/test_git_sync_scheduler.py]: Let's either have a dummy value or remove this one.

File: helm_tests/other/test_git_sync_scheduler.py [hunk 7]
```diff
@@ -211,6 +211,45 @@ def test_validate_if_ssh_params_are_added(self):
             "secret": {"secretName": "ssh-secret", "defaultMode": 288},
         } in jmespath.search("spec.template.spec.volumes", docs[0])
 
+    def test_validate_if_ssh_params_are_added_with_git_ssh_key(self):
+        docs = render_chart(
+            values={
+                "dags": {
+                    "gitSync": {
+                        "enabled": True,
+                        "containerName": "git-sync-test",
```
Comment (jedcunningham) [inline helm_tests/other/test_git_sync_scheduler.py]:
```suggestion
```

Don't think we need to set this.

File: chart/templates/secrets/git-ssh-key-secret.yaml (hunk 3 above)
Comment (romsharon98) [reply chart/templates/secrets/git-ssh-key-secret.yaml]: We supposed to get here 64b string but we don't convert it now

File: helm_tests/other/test_git_sync_scheduler.py (hunk 5 above)
Comment (romsharon98) [reply helm_tests/other/test_git_sync_scheduler.py]: removed

File: helm_tests/other/test_git_sync_scheduler.py (hunk 6 above)
Comment (romsharon98) [reply helm_tests/other/test_git_sync_scheduler.py]: removed also the assertation

File: helm_tests/other/test_git_sync_scheduler.py (hunk 7 above)
Comment (romsharon98) [reply helm_tests/other/test_git_sync_scheduler.py]: removed 😄

File: chart/values.schema.json (hunk 4 above)
Comment (romsharon98) [reply chart/values.schema.json]: changed 👍

File: helm_tests/other/test_git_sync_scheduler.py (hunk 5 above)
Comment (jedcunningham) [reply helm_tests/other/test_git_sync_scheduler.py]: It's still here.

File: chart/templates/secrets/git-ssh-key-secret.yaml (hunk 3 above)
Comment (jedcunningham) [reply chart/templates/secrets/git-ssh-key-secret.yaml]: Let me rephrase. Today, afaik, all our secret values are not base64 encoded in the values file. Is there a reason the sshKey needs to be encoded in the values file (say, newlines or something), or can this work (completely untested):

```
dags:
  gitSync:
    sshKey: |-
      -----BEGIN OPENSSH PRIVATE KEY-----
      ...
      -----END OPENSSH PRIVATE KEY-----
```

File: helm_tests/other/test_git_sync_scheduler.py (hunk 5 above)
Comment (romsharon98) [reply helm_tests/other/test_git_sync_scheduler.py]: right, now removed in all the tests I added.

File: chart/templates/secrets/git-ssh-key-secret.yaml (hunk 3 above)
Comment (romsharon98) [reply chart/templates/secrets/git-ssh-key-secret.yaml]: This will work.
So the value not have to be base64 :)

**Review by eladkal — APPROVED** [review]
LGTM

File: chart/templates/secrets/git-ssh-key-secret.yaml (hunk 3 above)
Comment (jedcunningham) [reply chart/templates/secrets/git-ssh-key-secret.yaml]: Unresolving this. Looks like we are still taking a base64 value?

Comment (romsharon98) [reply chart/templates/secrets/git-ssh-key-secret.yaml]: No we dont 😟

**Review by amoghrajesh — APPROVED** [review]
LGTM

**Comment by potiuk** [comment]
@jedcunningham :) ? Waits for your approval..

File: chart/templates/secrets/git-ssh-key-secret.yaml (hunk 3 above)
Comment (jedcunningham) [reply chart/templates/secrets/git-ssh-key-secret.yaml]: Here is what I see. In the secret:

```
data:
  gitSshKey: {{ .Values.dags.gitSync.sshKey | quote }}
```

And in the values file we have:

```
    # Or set sshKeySecret with your key
    # sshKey: <base64-converted-ssh-private-key>
```

That's why I say we are still expecting the user to put in a base64'd value - Secrets `data` takes base64 values, and it's documented as such in the values file.

Comment (jedcunningham) [reply chart/templates/secrets/git-ssh-key-secret.yaml]: I'm just saying it's inconsistent with the other secret values we gather from the user. If there isn't a technical reason to have them b64 it before hand, we should probably be consistent and take the raw value.

Comment (romsharon98) [reply chart/templates/secrets/git-ssh-key-secret.yaml]: I changed it, they don't have to be b64 before hand.

Comment (romsharon98) [reply chart/templates/secrets/git-ssh-key-secret.yaml]: https://github.com/apache/airflow/pull/39936#discussion_r1643492238

> Here is what I see. In the secret:
> 
> ```
> data:
>   gitSshKey: {{ .Values.dags.gitSync.sshKey | quote }}
> ```
> 
> And in the values file we have:
> 
> ```
>     # Or set sshKeySecret with your key
>     # sshKey: <base64-converted-ssh-private-key>
> ```
> 
> That's why I say we are still expecting the user to put in a base64'd value - Secrets `data` takes base64 values, and it's documented as such in the values file.

I totally missed this comment.
I got these comment from the doc here:
https://github.com/apache/airflow/blob/1f7be6f63e48d1861f913f2f5449a7a9e1bfd1fb/docs/helm-chart/manage-dags-files.rst?plain=1#L263
But I think it can get a raw value, therefore I removed the comment says it's need to be base64.

File: chart/values.schema.json [hunk 8]
```diff
@@ -7409,6 +7409,14 @@
                             ],
                             "default": null
                         },
+                        "sshKey": {
+                            "description": "Base64 converted ssh private key",
```
Comment (jedcunningham) [inline chart/values.schema.json]: This need to be updated.

File: chart/values.yaml [hunk 9]
```diff
@@ -2505,6 +2505,9 @@ dags:
     # and specify the name of the secret below
     # sshKeySecret: airflow-ssh-secret
     #
+    # Or set sshKeySecret with your key
+    # sshKey: <base64-converted-ssh-private-key>
```
Comment (jedcunningham) [inline chart/values.yaml]: And this. Probably worth putting an example with an empty key, since this is a multi-line deal.

Comment (romsharon98) [reply chart/values.yaml]: let me know if you ment like what I changed

File: chart/values.yaml [hunk 10]
```diff
@@ -2505,6 +2505,9 @@ dags:
     # and specify the name of the secret below
     # sshKeySecret: airflow-ssh-secret
     #
+    # Or set sshKeySecret with your key
+    # sshKey: ~
```
Comment (jedcunningham) [inline chart/values.yaml]:
```suggestion
    # Or set sshKey with your key
    # sshKey: |-
    #   -----BEGIN OPENSSH PRIVATE KEY-----
    #   ...
    #   -----END OPENSSH PRIVATE KEY-----
```

This is more what I was thinking.

Comment (romsharon98) [reply chart/values.yaml]: seems like this test does not allow it  "Detect if private key is added to the repository"

Comment (jedcunningham) [reply chart/values.yaml]: Maybe mangle the text a bit so the test doesn't hit on it, but it's (hopefully) obvious to a human what should be there?

```
-----BEGIN {YOUR-PRIVATE-KEY}-----
...
-----END {YOUR-PRIVATE KEY}-----
```

**Review by jedcunningham — APPROVED** [review]

**Comment by romsharon98** [comment]
@jedcunningham  , thank you for your detailed work on this PR—it’s greatly appreciated!
